### PR TITLE
store/tikv: update region error handling.

### PR DESCRIPTION
**Pull request by disksing**
Extract small functions and share region error handling logic between kv/coprocessor.

cc @coocood @tiancaiamao

### CODE REVIEW

File: store/tikv/region_request.go [hunk 1]
```diff
@@ -43,56 +59,125 @@ func sendKVReq(regionCache *RegionCache, rpcClient Client, bo *Backoffer, req *k
...
+
+func (s *RegionRequestSender) onSendFail(regionID RegionVerID, ctx *kvrpcpb.Context, err error) error {
+	s.regionCache.NextPeer(regionID)
+	err = s.bo.Backoff(boTiKVRPC, errors.Errorf("send tikv request error: %v, ctx: %s, try next peer later", err, ctx))
+	return errors.Trace(err)
+}
+
+func (s *RegionRequestSender) onRegionError(region *Region, regionErr *errorpb.Error) (retry bool, err error) {
+	reportRegionError(regionErr)
+
```
Comment (zimulala) [inline store/tikv/region_request.go]: Remove this useless line.

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -422,57 +423,23 @@ func (it *copIterator) handleTask(bo *Backoffer, task *copTask) (*coprocessor.Re
...
-				err = it.store.regionCache.OnRegionStale(task.region, staleEpoch.NewRegions)
-				if err != nil {
-					return nil, errors.Trace(err)
-				}
-			} else {
-				log.Warnf("tikv reports region error: %s, ctx: %s, retry later", e, req.Context)
-				it.store.regionCache.DropRegion(task.region.VerID())
-			}
-			err = bo.Backoff(boRegionMiss, errors.Errorf("regionError: %s", e))
+		if regionErr := resp.GetRegionError(); regionErr != nil {
```
Comment (shenli) [inline store/tikv/coprocessor.go]: If ServerIsBusy, we do not need to rebuild the task.

File: store/tikv/region_request.go [hunk 3]
```diff
@@ -18,22 +18,38 @@ import (
...
 	"github.com/pingcap/kvproto/pkg/kvrpcpb"
 )
 
-// sendKVReq sends req to tikv server. It will retry internally to find the right
-// region leader if i) fails to establish a connection to server or ii) server
-// returns `NotLeader`.
-func sendKVReq(regionCache *RegionCache, rpcClient Client, bo *Backoffer, req *kvrpcpb.Request, regionID RegionVerID, timeout time.Duration) (*kvrpcpb.Response, error) {
+// RegionRequestSender sends KV/Cop requests to tikv server.
+// It handles network errors and some region errors internally when it is safe to retry.
+type RegionRequestSender struct {
```
Comment (shenli) [inline store/tikv/region_request.go]: Could you add a few document for the sending and error handling logic?

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (disksing) [inline store/tikv/coprocessor.go]: `ServerIsBusy` is handled and retry inside the sender.

**Comment by disksing** [comment]
PTAL @zimulala @shenli

**Comment by zimulala** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**